### PR TITLE
link monitor

### PR DESCRIPTION
**Pull request by real-danm**
Adds link monitoring to the existing churn logic. 
Churn stays for scenarios where link monitor isn't ready or can't connect.

I used exponential moving average for the link monitor data points to ensure we were reactive to new datapoints. We can tweak the alpha value as needed to be more/less reactive to it.

```go test -v ./link_monitor_test.go ./link_monitor.go
=== RUN   TestNewLinkMonitorConfig
--- PASS: TestNewLinkMonitorConfig (0.00s)
=== RUN   TestNewLinkMonitor
--- PASS: TestNewLinkMonitor (0.00s)
=== RUN   TestDisabledLinkMonitor
time="2023-10-18T17:30:40-06:00" level=info msg="not enough data points for region" region=region1 vpnIp=10.1.0.1
--- PASS: TestDisabledLinkMonitor (0.00s)
=== RUN   TestLinkMonitorCheckEmpty
time="2023-10-18T17:30:40-06:00" level=info msg="not enough data points for region" region=region1 vpnIp=10.1.0.1
time="2023-10-18T17:30:40-06:00" level=info msg="not enough data points for region" region=region1 vpnIp=10.1.0.1
--- PASS: TestLinkMonitorCheckEmpty (0.00s)
=== RUN   TestLinkMonitorAttempt
time="2023-10-18T17:30:40-06:00" level=info msg="not enough data points for region" region=region2 vpnIp=10.2.0.1
--- PASS: TestLinkMonitorAttempt (0.00s)
=== RUN   TestCheckAndAttemptWithDeadEndpoint
time="2023-10-18T17:30:40-06:00" level=info msg="not enough data points for region" region=region1 vpnIp=10.1.0.1
time="2023-10-18T17:30:40-06:00" level=info msg="not enough data points for region" region=region1 vpnIp=10.1.0.1
--- PASS: TestCheckAndAttemptWithDeadEndpoint (0.00s)
=== RUN   TestFailureDueToRtt
time="2023-10-18T17:30:40-06:00" level=info msg="not enough data points for region" region= vpnIp=10.1.0.1
time="2023-10-18T17:30:40-06:00" level=info msg="not enough data points for region" region= vpnIp=10.1.0.1
time="2023-10-18T17:30:40-06:00" level=error msg="failed to check regional statsfailed to find regional stats for vpnIp 10.15.0.1"
time="2023-10-18T17:30:40-06:00" level=warning msg="region has high rtt or jitter" jitter=0 region= rtt=1e+06 vpnIp=10.1.0.1
time="2023-10-18T17:30:40-06:00" level=error msg="failed to check regional statsfailed to find regional stats for vpnIp 10.15.0.1"
time="2023-10-18T17:30:40-06:00" level=error msg="failed to check regional statsfailed to find regional stats for vpnIp 10.15.0.1"
time="2023-10-18T17:30:40-06:00" level=error msg="failed to check regional statsfailed to find regional stats for vpnIp 10.15.0.1"
--- PASS: TestFailureDueToRtt (0.00s)
=== RUN   TestFailureDueToJitter
time="2023-10-18T17:30:40-06:00" level=info msg="not enough data points for region" region= vpnIp=10.1.0.1
time="2023-10-18T17:30:40-06:00" level=info msg="not enough data points for region" region= vpnIp=10.1.0.1
time="2023-10-18T17:30:40-06:00" level=error msg="failed to check regional statsfailed to find regional stats for vpnIp 10.15.0.1"
time="2023-10-18T17:30:40-06:00" level=warning msg="region has high rtt or jitter" jitter=75000 region= rtt=0 vpnIp=10.1.0.1
time="2023-10-18T17:30:40-06:00" level=error msg="failed to check regional statsfailed to find regional stats for vpnIp 10.15.0.1"
time="2023-10-18T17:30:40-06:00" level=error msg="failed to check regional statsfailed to find regional stats for vpnIp 10.15.0.1"
time="2023-10-18T17:30:40-06:00" level=error msg="failed to check regional statsfailed to find regional stats for vpnIp 10.15.0.1"
--- PASS: TestFailureDueToJitter (0.00s)
PASS
ok  	command-line-arguments	0.233s```

### CODE REVIEW

File: handshake_manager.go [hunk 1]
```diff
@@ -38,7 +38,8 @@ type HandshakeConfig struct {
 	triggerBuffer int
 	useRelays     bool
 
-	messageMetrics *MessageMetrics
+	linkMonitorConfig map[interface{}]interface{}
```
Comment (matkam) [inline handshake_manager.go:41]: can this be a struct with defined fields?

Comment (real-danm) [reply handshake_manager.go:41]: Unfortunately not due to the way the nebula config works, have to grab the yaml with `linkMonitorConfig: c.GetMap("link_monitor", map[interface{}]interface{}{}),`
from https://github.com/real-danm/nebula/blob/master/config/config.go#L214

Comment (matkam) [reply handshake_manager.go:41]: got it. how about still making it into a struct earlier on, and populating the fields with `GetBool()`, `GetString()`, etc?

**Review by matkam — APPROVED** [review]
lgtm! nice work

**Comment by real-danm** [comment]
will re-open this later

**Pull request closed without merging**